api/uniter/unit_test.go

```diff
@@ -684,6 +684,86 @@ func (s *unitSuite) TestWatchActionNotificationsMoreResults(c *gc.C) {
 	c.Assert(err.Error(), gc.Equals, "expected 1 result, got 2")
 }
 
+func (s *unitSuite) TestWatchUpgradeSeriesNotifications(c *gc.C) {
+	watcher, err := s.apiUnit.WatchUpgradeSeriesNotifications()
+	c.Assert(err, jc.ErrorIsNil)
+
+	notifyWatcher := watchertest.NewNotifyWatcherC(c, watcher, s.BackingState.StartSync)
+	defer notifyWatcher.AssertStops()
+
+	// Why do I have two initial events?!?
+	_, ok := <-notifyWatcher.Watcher.Changes()
+	notifyWatcher.Assert(ok, jc.IsTrue)
+
+	notifyWatcher.AssertOneChange()
+
+	s.CreateUpgradeSeriesLock(c)
+
+	// Expect a notification that the document was created (i.e. a lock was placed)
+	notifyWatcher.AssertOneChange()
+
+	err = s.wordpressMachine.RemoveUpgradeSeriesLock()
+	c.Assert(err, jc.ErrorIsNil)
+
+	// A notification that the document was removed (i.e. the lock was released)
+	notifyWatcher.AssertOneChange()
+}
+
+func (s *unitSuite) TestUpgradeSeriesStatus(c *gc.C) {
+	// First we create the prepare lock
+	s.CreateUpgradeSeriesLock(c)
+
+	// Then we check to see the status of our upgrade. We note that creating
+	// the lock essentially kicks off an upgrade for the perspective of
+	// assigned units.
+	status, err := s.apiUnit.UpgradeSeriesStatus()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(status, gc.Equals, params.UnitStarted)
+}
+
+func (s *unitSuite) TestSetUpgradeSeriesStatus(c *gc.C) {
+	// First we create the prepare lock or the required state will not exists
+	s.CreateUpgradeSeriesLock(c)
+
+	err := s.apiUnit.SetUpgradeSeriesStatus(params.UnitStarted)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Check to see that the upgrade status has been set appropriately
+	status, err := s.apiUnit.UpgradeSeriesStatus()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(status, gc.Equals, params.UnitStarted)
+}
+
+func (s *unitSuite) TestSetUpgradeSeriesStatusShouldOnlySetSpecifiedUnit(c *gc.C) {
+	// add another unit
+	unit2, err := s.wordpressApplication.AddUnit(state.AddUnitParams{})
+	c.Assert(err, jc.ErrorIsNil)
+
+	err = unit2.AssignToMachine(s.wordpressMachine)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// Creating a lock for the machine transitions all units to started state
+	s.CreateUpgradeSeriesLock(c, unit2.Name())
+
+	// Complete one unit
+	_, err = unit2.SetUpgradeSeriesStatus(params.UnitCompleted)
+	c.Assert(err, jc.ErrorIsNil)
+
+	// The other unit should still be in the started state
+	status, err := s.wordpressUnit.UpgradeSeriesStatus()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(status, gc.Equals, params.UnitStarted)
+}
+
+func (s *unitSuite) CreateUpgradeSeriesLock(c *gc.C, additionalUnits ...string) {
+	unitNames := additionalUnits
+	unitNames = append(unitNames, s.wordpressUnit.Name())
+	series := "trust"
```

Comment:
Are we going to add validation that the series mentioned is one that juju knows about?

Reply:
There is validation at the CLI level. Will add validation at the DB level too.

Reply:
It would probably make sense to do the validation at the API server level rather than the DB level.

Reply:
we do validate the series via an api call, before calling prepare, as the machine lock is created... perhaps it can be validated again when creating the machine lock document?